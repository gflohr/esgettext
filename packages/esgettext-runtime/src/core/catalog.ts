export interface CatalogEntries {
	[key: string]: Array<string>;
}

export interface Catalog {
	major: number;
	minor: number;
	pluralFunction: Function;
	entries: CatalogEntries;
}
