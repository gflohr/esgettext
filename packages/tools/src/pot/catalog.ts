import { POTEntry } from './entry';

interface Cache {
	// msgid
	[key: string]: {
		// msgctxt
		[key: string]: POTEntry;
	};
}

interface CatalogProperties {
	package?: string;
	version?: string;
	msgidBugsAddress?: string;
	foreignUser?: boolean;
	date?: string;
	fromCode?: string;
	sortOutput?: boolean;
	sortByFile?: boolean;
}

interface RenderOptions {
	width?: number;
}

/**
 * An entire message catalog that is subsequently filled.
 */
export class Catalog {
	private readonly cache: Cache = {};
	private readonly entries: Array<POTEntry>;

	constructor(private readonly properties: CatalogProperties = {}) {
		if (typeof properties.package === 'undefined') {
			properties.package = 'PACKAGE';
		}
		if (typeof properties.version === 'undefined') {
			properties.version = 'VERSION';
		}
		if (typeof properties.msgidBugsAddress === 'undefined') {
			properties.msgidBugsAddress = 'MSGID_BUGS_ADDRESS';
		}
		// GNU xgettext does not set the charset in the PO header but we do.
		if (typeof properties.fromCode === 'undefined') {
			properties.fromCode = 'CHARSET';
		}
		if (typeof properties.date === 'undefined') {
			const now = new Date();

			// Avoid if/else, so we do not spoil our test coverage. :)
			let year = now.getFullYear().toString();
			year = '0'.repeat(4 - year.length) + year;
			let month = now.getMonth().toString();
			month = '0'.repeat(2 - month.length) + month;
			let mday = now.getDate().toString();
			mday = '0'.repeat(2 - mday.length) + mday;
			let hour = now.getHours().toString();
			hour = '0'.repeat(2 - hour.length) + hour;
			let minutes = now.getMinutes().toString();
			minutes = '0'.repeat(2 - minutes.length) + minutes;

			// Do not depend on the timezone for test coverage.
			let offset = now.getTimezoneOffset();
			const sign = ['+', '-'][Math.sign(offset) + 1];
			offset = Math.abs(offset);

			let offsetHours = Math.floor(offset / 60).toString();
			offsetHours = '0'.repeat(2 - offsetHours.length) + offsetHours;
			let offsetMinutes = (offset % 60).toString();
			offsetMinutes = '0'.repeat(2 - offsetMinutes.length) + offsetMinutes;

			properties.date =
				[year, month, mday].join('-') +
				` ${hour}:${minutes}` +
				`${sign}${offsetHours}${offsetMinutes}`;
		}

		const pkg = properties.package.replace(/\n/g, '\\n');
		const version = properties.version.replace(/\n/g, '\\n');
		const msgidBugsAddress = properties.msgidBugsAddress.replace(/\n/g, '\\n');

		const header = `Project-Id-Version: ${pkg} ${version}
Report-Msgid-Bugs-To: ${msgidBugsAddress}
POT-Creation-Date: ${properties.date}
PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE
Last-Translator: FULL NAME <EMAIL@ADDRESS>
Language-Team: LANGUAGE <LL@li.org>
Language:
MIME-Version: 1.0
Content-Type: text/plain; charset=${properties.fromCode}
Content-Transfer-Encoding: 8bit
`;

		let comment;
		if (properties.foreignUser) {
			comment = `SOME DESCRIPTIVE TITLE
This file is put in the public domain.
FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
`;
		} else {
			comment = `SOME DESCRIPTIVE TITLE
Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
This file is distributed under the same license as the PACKAGE package.
FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
`;
		}

		this.entries = new Array<POTEntry>();
		const headerEntry = new POTEntry({
			msgid: '',
			msgstr: header,
			flags: ['fuzzy'],
			translatorComments: [comment],
			noWarnings: true,
		});
		this.addEntry(headerEntry);
	}

	/**
	 * Render the entire catalog.
	 *
	 * @param renderOptions - options for rendering
	 *
	 * The following options are supported:
	 *
	 * width: wrap lines to that page width (default: 76)
	 *
	 * @returns the catalog rendered as a po(t) file
	 */
	toString(options?: RenderOptions): string {
		const width =
			options && typeof options.width !== 'undefined' ? options.width : 76;

		if (this.properties.sortOutput) {
			return this.entries
				.sort((a, b) => {
					const cmp = a.properties.msgid.localeCompare(b.properties.msgid);
					if (cmp) {
						return cmp;
					}
					const actx = a.properties.msgctxt;
					const bctx = b.properties.msgctxt;

					if (typeof actx === 'undefined') {
						if (typeof bctx === 'undefined') {
							return 0;
						} else {
							return -1;
						}
					} else {
						if (typeof bctx === 'undefined') {
							return +1;
						} else {
							return actx.localeCompare(bctx);
						}
					}
				})
				.map((entry) => entry.toString(width))
				.join('\n');
		} else {
			return this.entries.map((entry) => entry.toString(width)).join('\n');
		}
	}

	/**
	 * Add an entry to the catalog. If an entry with the same msgid
	 * and msgctxt already exists, the entry is merged into
	 * the existing one instead.
	 *
	 * @param entry - the `POTEntry` to add
	 */
	addEntry(entry: POTEntry): void {
		const msgid = entry.properties.msgid;
		const msgctxt = entry.properties.msgctxt;

		if (!Object.prototype.hasOwnProperty.call(this.cache, msgid)) {
			this.cache[msgid] = {};
		}
		if (!Object.prototype.hasOwnProperty.call(this.cache[msgid], msgctxt)) {
			this.cache[msgid][msgctxt] = entry;
			this.entries.push(entry);
		} else {
			this.cache[msgid][msgctxt].merge(entry);
		}
	}

	/**
	 * Copy a catalog with other options. This is for testing only.
	 *
	 * @returns a deep copy of the catalog.
	 */
	copy(properties?: CatalogProperties): Catalog {
		const other = new Catalog(properties);

		this.entries.forEach((entry) => other.addEntry(entry));

		return other;
	}
}
