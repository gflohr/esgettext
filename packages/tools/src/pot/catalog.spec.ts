import { Catalog } from './catalog';
import { POTEntry } from './entry';

const date = '2020-04-23 08:50+0300';

describe('translation catalog', () => {
	describe('initialization', () => {
		it('should be initialized with zero configuration', () => {
			const catalog = new Catalog();
			expect(catalog.toString()).toMatch(/Content-Type/);
		});

		it('should be initialized with default values', () => {
			const catalog = new Catalog({ date });
			expect(catalog.toString()).toMatchSnapshot();
		});

		it('should honor the foreign-user option', () => {
			const catalog = new Catalog({ date, foreignUser: true });
			expect(catalog.toString()).toMatchSnapshot();
		});

		it('should honor the package option', () => {
			const catalog = new Catalog({ date, package: 'foobar' });
			expect(catalog.toString()).toMatchSnapshot();
		});

		it('should honor the version option', () => {
			const catalog = new Catalog({ date, version: '23.4.89' });
			expect(catalog.toString()).toMatchSnapshot();
		});

		it('should honor the msgid-bugs-address option', () => {
			const catalog = new Catalog({
				date,
				msgidBugsAddress: 'me@example.com',
			});
			expect(catalog.toString()).toMatchSnapshot();
		});

		it('should honor the from-code option', () => {
			const catalog = new Catalog({ date, fromCode: 'utf-8' });
			expect(catalog.toString()).toMatchSnapshot();
		});
	});

	describe('conflicting flags', () => {
		/* eslint-disable-next-line @typescript-eslint/no-explicit-any */
		let warner: jest.SpyInstance<void, [any?, ...any[]]>;

		beforeEach(() => {
			warner = jest.spyOn(global.console, 'warn').mockImplementation(() => {
				/* ignore */
			});
		});

		afterEach(() => jest.restoreAllMocks());

		it('should warn about conflicting flags', () => {
			const catalog = new Catalog();

			catalog.addEntry(
				new POTEntry({
					msgid: 'hello',
					flags: ['no-mercy', 'perl-brace-format'],
					references: ['first.js:10'],
				}),
			);

			catalog.addEntry(
				new POTEntry({
					msgid: 'hello',
					flags: ['no-mercy'],
					references: ['second.js:10'],
				}),
			);

			catalog.addEntry(
				new POTEntry({
					msgid: 'hello',
					flags: ['mercy'],
					references: ['third.js:10'],
				}),
			);
			expect(warner).toHaveBeenCalledTimes(4);

			catalog.addEntry(
				new POTEntry({
					msgid: 'hello',
					flags: ['no-perl-brace-format'],
					references: ['third.js:10'],
				}),
			);
			expect(warner).toHaveBeenCalledTimes(9);
		});
	});

	describe('continuous filling', () => {
		const catalog = new Catalog({ date });

		it('should add an entry', () => {
			catalog.addEntry(
				new POTEntry({
					msgid: 'Mike',
					references: ['src/zulu.ts:2304'],
				}),
			);
			expect(catalog.toString()).toMatchSnapshot();
		});

		it('should add a second entry', () => {
			catalog.addEntry(
				new POTEntry({
					msgid: 'November',
					references: ['src/yankee.ts:42'],
				}),
			);
			expect(catalog.toString()).toMatchSnapshot();
		});

		it('should merge the third entry', () => {
			catalog.addEntry(
				new POTEntry({
					msgid: 'Mike',
					references: ['src/xray.ts:1303'],
				}),
			);
			expect(catalog.toString()).toMatchSnapshot();
		});

		// This can later be used to test sorting by file.
		it('should support a string at many locations', () => {
			catalog.addEntry(
				new POTEntry({
					msgid: 'Lima',
					references: ['whiskey.ts:2'],
				}),
			);
			catalog.addEntry(
				new POTEntry({
					msgid: 'Oscar',
					references: ['whiskey.ts:2'],
				}),
			);
			catalog.addEntry(
				new POTEntry({
					msgid: 'Lima',
					references: ['victor.ts:3'],
				}),
			);
			catalog.addEntry(
				new POTEntry({
					msgid: 'Oscar',
					references: ['victor.ts:3'],
				}),
			);
			// When sorting by file, the Oscar entry should appear before
			// the Lima entry. Actually that case could not happen but ...
			catalog.addEntry(
				new POTEntry({
					msgid: 'Lima',
					references: ['uniform.ts:4'],
				}),
			);
			catalog.addEntry(
				new POTEntry({
					msgid: 'Oscar',
					references: ['uniform.ts:3'],
				}),
			);
			catalog.addEntry(
				new POTEntry({
					msgid: 'Lima',
					references: ['tango.ts:25'],
				}),
			);
			catalog.addEntry(
				new POTEntry({
					msgid: 'Oscar',
					references: ['tango.ts:25'],
				}),
			);
			expect(catalog.toString()).toMatchSnapshot();
		});
		it('should also merge flags and comments', () => {
			catalog.addEntry(
				new POTEntry({
					msgid: 'Mike',
					translatorComments: ['This is difficult.'],
					automatic: ['TRANSLATORS: Buy me a milk shake!'],
					flags: ['no-c-format'],
					references: ['sierra.ts:27'],
				}),
			);
			expect(catalog.toString()).toMatchSnapshot();

			catalog.addEntry(
				new POTEntry({
					msgid: 'Mike',
					translatorComments: ['Not really.'],
					automatic: ['TRANSLATORS: Drop me a line!'],
					flags: ['perl-brace-format'],
					references: ['romeo.ts:85'],
				}),
			);
			expect(catalog.toString()).toMatchSnapshot();

			// We need entries with a message context.
			catalog.addEntry(
				new POTEntry({
					msgid: 'Mike',
					msgctxt: 'phonetic alphabet',
					references: ['juliet.ts:85'],
				}),
			);
			catalog.addEntry(
				new POTEntry({
					msgid: 'Mike',
					msgctxt: 'audio',
					references: ['golf.ts:85'],
				}),
			);

			// And entries with one more ref than another.
			catalog.addEntry(
				new POTEntry({
					msgid: 'Foxtrot',
					references: ['src/yankee.ts:42', 'src/zulu.ts:2304'],
				}),
			);
		});

		it('should sort output on demand', () => {
			const copy = catalog.copy({ sortOutput: true, date });

			expect(copy.toString()).toMatchSnapshot();
		});

		it('should sort by file on demand', () => {
			const copy = catalog.copy({ sortByFile: true, date });

			expect(copy.toString()).toMatchSnapshot();
		});
	});
});
