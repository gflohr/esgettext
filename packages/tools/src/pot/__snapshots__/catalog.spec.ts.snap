// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`translation catalog continuous filling should add a second entry 1`] = `
"# SOME DESCRIPTIVE TITLE
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
# 
#, fuzzy
msgid \\"\\"
msgstr \\"\\"
\\"Project-Id-Version: PACKAGE VERSION\\\\n\\"
\\"Report-Msgid-Bugs-To: MSGID_BUGS_ADDRESS\\\\n\\"
\\"POT-Creation-Date: 2020-04-23 08:50+0300\\\\n\\"
\\"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\\\\n\\"
\\"Last-Translator: FULL NAME <EMAIL@ADDRESS>\\\\n\\"
\\"Language-Team: LANGUAGE <LL@li.org>\\\\n\\"
\\"Language:\\\\n\\"
\\"MIME-Version: 1.0\\\\n\\"
\\"Content-Type: text/plain; charset=CHARSET\\\\n\\"
\\"Content-Transfer-Encoding: 8bit\\\\n\\"

#: src/zulu.ts:2304
msgid \\"Mike\\"
msgstr \\"\\"

#: src/yankee.ts:42
msgid \\"November\\"
msgstr \\"\\"
"
`;

exports[`translation catalog continuous filling should add an entry 1`] = `
"# SOME DESCRIPTIVE TITLE
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
# 
#, fuzzy
msgid \\"\\"
msgstr \\"\\"
\\"Project-Id-Version: PACKAGE VERSION\\\\n\\"
\\"Report-Msgid-Bugs-To: MSGID_BUGS_ADDRESS\\\\n\\"
\\"POT-Creation-Date: 2020-04-23 08:50+0300\\\\n\\"
\\"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\\\\n\\"
\\"Last-Translator: FULL NAME <EMAIL@ADDRESS>\\\\n\\"
\\"Language-Team: LANGUAGE <LL@li.org>\\\\n\\"
\\"Language:\\\\n\\"
\\"MIME-Version: 1.0\\\\n\\"
\\"Content-Type: text/plain; charset=CHARSET\\\\n\\"
\\"Content-Transfer-Encoding: 8bit\\\\n\\"

#: src/zulu.ts:2304
msgid \\"Mike\\"
msgstr \\"\\"
"
`;

exports[`translation catalog continuous filling should also merge flags and comments 1`] = `
"# SOME DESCRIPTIVE TITLE
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
# 
#, fuzzy
msgid \\"\\"
msgstr \\"\\"
\\"Project-Id-Version: PACKAGE VERSION\\\\n\\"
\\"Report-Msgid-Bugs-To: MSGID_BUGS_ADDRESS\\\\n\\"
\\"POT-Creation-Date: 2020-04-23 08:50+0300\\\\n\\"
\\"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\\\\n\\"
\\"Last-Translator: FULL NAME <EMAIL@ADDRESS>\\\\n\\"
\\"Language-Team: LANGUAGE <LL@li.org>\\\\n\\"
\\"Language:\\\\n\\"
\\"MIME-Version: 1.0\\\\n\\"
\\"Content-Type: text/plain; charset=CHARSET\\\\n\\"
\\"Content-Transfer-Encoding: 8bit\\\\n\\"

# This is difficult.
#. TRANSLATORS: Buy me a milk shake!
#: src/zulu.ts:2304 src/xray.ts:1303 sierra.ts:27
#, no-c-format
msgid \\"Mike\\"
msgstr \\"\\"

#: src/yankee.ts:42
msgid \\"November\\"
msgstr \\"\\"

#: whiskey.ts:2 victor.ts:3 uniform.ts:4 tango.ts:25
msgid \\"Lima\\"
msgstr \\"\\"

#: whiskey.ts:2 victor.ts:3 uniform.ts:3 tango.ts:25
msgid \\"Oscar\\"
msgstr \\"\\"
"
`;

exports[`translation catalog continuous filling should also merge flags and comments 2`] = `
"# SOME DESCRIPTIVE TITLE
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
# 
#, fuzzy
msgid \\"\\"
msgstr \\"\\"
\\"Project-Id-Version: PACKAGE VERSION\\\\n\\"
\\"Report-Msgid-Bugs-To: MSGID_BUGS_ADDRESS\\\\n\\"
\\"POT-Creation-Date: 2020-04-23 08:50+0300\\\\n\\"
\\"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\\\\n\\"
\\"Last-Translator: FULL NAME <EMAIL@ADDRESS>\\\\n\\"
\\"Language-Team: LANGUAGE <LL@li.org>\\\\n\\"
\\"Language:\\\\n\\"
\\"MIME-Version: 1.0\\\\n\\"
\\"Content-Type: text/plain; charset=CHARSET\\\\n\\"
\\"Content-Transfer-Encoding: 8bit\\\\n\\"

# This is difficult.
# Not really.
#. TRANSLATORS: Buy me a milk shake!
#. TRANSLATORS: Drop me a line!
#: src/zulu.ts:2304 src/xray.ts:1303 sierra.ts:27 romeo.ts:85
#, no-c-format, perl-brace-format
msgid \\"Mike\\"
msgstr \\"\\"

#: src/yankee.ts:42
msgid \\"November\\"
msgstr \\"\\"

#: whiskey.ts:2 victor.ts:3 uniform.ts:4 tango.ts:25
msgid \\"Lima\\"
msgstr \\"\\"

#: whiskey.ts:2 victor.ts:3 uniform.ts:3 tango.ts:25
msgid \\"Oscar\\"
msgstr \\"\\"
"
`;

exports[`translation catalog continuous filling should merge the third entry 1`] = `
"# SOME DESCRIPTIVE TITLE
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
# 
#, fuzzy
msgid \\"\\"
msgstr \\"\\"
\\"Project-Id-Version: PACKAGE VERSION\\\\n\\"
\\"Report-Msgid-Bugs-To: MSGID_BUGS_ADDRESS\\\\n\\"
\\"POT-Creation-Date: 2020-04-23 08:50+0300\\\\n\\"
\\"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\\\\n\\"
\\"Last-Translator: FULL NAME <EMAIL@ADDRESS>\\\\n\\"
\\"Language-Team: LANGUAGE <LL@li.org>\\\\n\\"
\\"Language:\\\\n\\"
\\"MIME-Version: 1.0\\\\n\\"
\\"Content-Type: text/plain; charset=CHARSET\\\\n\\"
\\"Content-Transfer-Encoding: 8bit\\\\n\\"

#: src/zulu.ts:2304 src/xray.ts:1303
msgid \\"Mike\\"
msgstr \\"\\"

#: src/yankee.ts:42
msgid \\"November\\"
msgstr \\"\\"
"
`;

exports[`translation catalog continuous filling should sort by file on demand 1`] = `
"# SOME DESCRIPTIVE TITLE
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
# 
# SOME DESCRIPTIVE TITLE
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
# 
# SOME DESCRIPTIVE TITLE
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
# 
#, fuzzy
msgid \\"\\"
msgstr \\"\\"
\\"Project-Id-Version: PACKAGE VERSION\\\\n\\"
\\"Report-Msgid-Bugs-To: MSGID_BUGS_ADDRESS\\\\n\\"
\\"POT-Creation-Date: 2020-04-23 08:50+0300\\\\n\\"
\\"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\\\\n\\"
\\"Last-Translator: FULL NAME <EMAIL@ADDRESS>\\\\n\\"
\\"Language-Team: LANGUAGE <LL@li.org>\\\\n\\"
\\"Language:\\\\n\\"
\\"MIME-Version: 1.0\\\\n\\"
\\"Content-Type: text/plain; charset=CHARSET\\\\n\\"
\\"Content-Transfer-Encoding: 8bit\\\\n\\"

#: golf.ts:85
msgctxt \\"audio\\"
msgid \\"Mike\\"
msgstr \\"\\"

#: juliet.ts:85
msgctxt \\"phonetic alphabet\\"
msgid \\"Mike\\"
msgstr \\"\\"

# This is difficult.
# Not really.
#. TRANSLATORS: Buy me a milk shake!
#. TRANSLATORS: Drop me a line!
#: romeo.ts:85 sierra.ts:27 src/xray.ts:1303 src/zulu.ts:2304
#, no-c-format, perl-brace-format
msgid \\"Mike\\"
msgstr \\"\\"

#: src/yankee.ts:42
msgid \\"November\\"
msgstr \\"\\"

#: src/yankee.ts:42 src/zulu.ts:2304
msgid \\"Foxtrot\\"
msgstr \\"\\"

#: tango.ts:25 uniform.ts:3 victor.ts:3 whiskey.ts:2
msgid \\"Oscar\\"
msgstr \\"\\"

#: tango.ts:25 uniform.ts:4 victor.ts:3 whiskey.ts:2
msgid \\"Lima\\"
msgstr \\"\\"
"
`;

exports[`translation catalog continuous filling should sort output on demand 1`] = `
"# SOME DESCRIPTIVE TITLE
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
# 
# SOME DESCRIPTIVE TITLE
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
# 
#, fuzzy
msgid \\"\\"
msgstr \\"\\"
\\"Project-Id-Version: PACKAGE VERSION\\\\n\\"
\\"Report-Msgid-Bugs-To: MSGID_BUGS_ADDRESS\\\\n\\"
\\"POT-Creation-Date: 2020-04-23 08:50+0300\\\\n\\"
\\"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\\\\n\\"
\\"Last-Translator: FULL NAME <EMAIL@ADDRESS>\\\\n\\"
\\"Language-Team: LANGUAGE <LL@li.org>\\\\n\\"
\\"Language:\\\\n\\"
\\"MIME-Version: 1.0\\\\n\\"
\\"Content-Type: text/plain; charset=CHARSET\\\\n\\"
\\"Content-Transfer-Encoding: 8bit\\\\n\\"

#: src/yankee.ts:42 src/zulu.ts:2304
msgid \\"Foxtrot\\"
msgstr \\"\\"

#: whiskey.ts:2 victor.ts:3 uniform.ts:4 tango.ts:25
msgid \\"Lima\\"
msgstr \\"\\"

# This is difficult.
# Not really.
#. TRANSLATORS: Buy me a milk shake!
#. TRANSLATORS: Drop me a line!
#: src/zulu.ts:2304 src/xray.ts:1303 sierra.ts:27 romeo.ts:85
#, no-c-format, perl-brace-format
msgid \\"Mike\\"
msgstr \\"\\"

#: golf.ts:85
msgctxt \\"audio\\"
msgid \\"Mike\\"
msgstr \\"\\"

#: juliet.ts:85
msgctxt \\"phonetic alphabet\\"
msgid \\"Mike\\"
msgstr \\"\\"

#: src/yankee.ts:42
msgid \\"November\\"
msgstr \\"\\"

#: whiskey.ts:2 victor.ts:3 uniform.ts:3 tango.ts:25
msgid \\"Oscar\\"
msgstr \\"\\"
"
`;

exports[`translation catalog continuous filling should support a string at many locations 1`] = `
"# SOME DESCRIPTIVE TITLE
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
# 
#, fuzzy
msgid \\"\\"
msgstr \\"\\"
\\"Project-Id-Version: PACKAGE VERSION\\\\n\\"
\\"Report-Msgid-Bugs-To: MSGID_BUGS_ADDRESS\\\\n\\"
\\"POT-Creation-Date: 2020-04-23 08:50+0300\\\\n\\"
\\"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\\\\n\\"
\\"Last-Translator: FULL NAME <EMAIL@ADDRESS>\\\\n\\"
\\"Language-Team: LANGUAGE <LL@li.org>\\\\n\\"
\\"Language:\\\\n\\"
\\"MIME-Version: 1.0\\\\n\\"
\\"Content-Type: text/plain; charset=CHARSET\\\\n\\"
\\"Content-Transfer-Encoding: 8bit\\\\n\\"

#: src/zulu.ts:2304 src/xray.ts:1303
msgid \\"Mike\\"
msgstr \\"\\"

#: src/yankee.ts:42
msgid \\"November\\"
msgstr \\"\\"

#: whiskey.ts:2 victor.ts:3 uniform.ts:4 tango.ts:25
msgid \\"Lima\\"
msgstr \\"\\"

#: whiskey.ts:2 victor.ts:3 uniform.ts:3 tango.ts:25
msgid \\"Oscar\\"
msgstr \\"\\"
"
`;

exports[`translation catalog initialization should be initialized with default values 1`] = `
"# SOME DESCRIPTIVE TITLE
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
# 
#, fuzzy
msgid \\"\\"
msgstr \\"\\"
\\"Project-Id-Version: PACKAGE VERSION\\\\n\\"
\\"Report-Msgid-Bugs-To: MSGID_BUGS_ADDRESS\\\\n\\"
\\"POT-Creation-Date: 2020-04-23 08:50+0300\\\\n\\"
\\"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\\\\n\\"
\\"Last-Translator: FULL NAME <EMAIL@ADDRESS>\\\\n\\"
\\"Language-Team: LANGUAGE <LL@li.org>\\\\n\\"
\\"Language:\\\\n\\"
\\"MIME-Version: 1.0\\\\n\\"
\\"Content-Type: text/plain; charset=CHARSET\\\\n\\"
\\"Content-Transfer-Encoding: 8bit\\\\n\\"
"
`;

exports[`translation catalog initialization should honor the foreign-user option 1`] = `
"# SOME DESCRIPTIVE TITLE
# This file is put in the public domain.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
# 
#, fuzzy
msgid \\"\\"
msgstr \\"\\"
\\"Project-Id-Version: PACKAGE VERSION\\\\n\\"
\\"Report-Msgid-Bugs-To: MSGID_BUGS_ADDRESS\\\\n\\"
\\"POT-Creation-Date: 2020-04-23 08:50+0300\\\\n\\"
\\"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\\\\n\\"
\\"Last-Translator: FULL NAME <EMAIL@ADDRESS>\\\\n\\"
\\"Language-Team: LANGUAGE <LL@li.org>\\\\n\\"
\\"Language:\\\\n\\"
\\"MIME-Version: 1.0\\\\n\\"
\\"Content-Type: text/plain; charset=CHARSET\\\\n\\"
\\"Content-Transfer-Encoding: 8bit\\\\n\\"
"
`;

exports[`translation catalog initialization should honor the from-code option 1`] = `
"# SOME DESCRIPTIVE TITLE
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
# 
#, fuzzy
msgid \\"\\"
msgstr \\"\\"
\\"Project-Id-Version: PACKAGE VERSION\\\\n\\"
\\"Report-Msgid-Bugs-To: MSGID_BUGS_ADDRESS\\\\n\\"
\\"POT-Creation-Date: 2020-04-23 08:50+0300\\\\n\\"
\\"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\\\\n\\"
\\"Last-Translator: FULL NAME <EMAIL@ADDRESS>\\\\n\\"
\\"Language-Team: LANGUAGE <LL@li.org>\\\\n\\"
\\"Language:\\\\n\\"
\\"MIME-Version: 1.0\\\\n\\"
\\"Content-Type: text/plain; charset=utf-8\\\\n\\"
\\"Content-Transfer-Encoding: 8bit\\\\n\\"
"
`;

exports[`translation catalog initialization should honor the msgid-bugs-address option 1`] = `
"# SOME DESCRIPTIVE TITLE
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
# 
#, fuzzy
msgid \\"\\"
msgstr \\"\\"
\\"Project-Id-Version: PACKAGE VERSION\\\\n\\"
\\"Report-Msgid-Bugs-To: me@example.com\\\\n\\"
\\"POT-Creation-Date: 2020-04-23 08:50+0300\\\\n\\"
\\"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\\\\n\\"
\\"Last-Translator: FULL NAME <EMAIL@ADDRESS>\\\\n\\"
\\"Language-Team: LANGUAGE <LL@li.org>\\\\n\\"
\\"Language:\\\\n\\"
\\"MIME-Version: 1.0\\\\n\\"
\\"Content-Type: text/plain; charset=CHARSET\\\\n\\"
\\"Content-Transfer-Encoding: 8bit\\\\n\\"
"
`;

exports[`translation catalog initialization should honor the package option 1`] = `
"# SOME DESCRIPTIVE TITLE
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
# 
#, fuzzy
msgid \\"\\"
msgstr \\"\\"
\\"Project-Id-Version: foobar VERSION\\\\n\\"
\\"Report-Msgid-Bugs-To: MSGID_BUGS_ADDRESS\\\\n\\"
\\"POT-Creation-Date: 2020-04-23 08:50+0300\\\\n\\"
\\"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\\\\n\\"
\\"Last-Translator: FULL NAME <EMAIL@ADDRESS>\\\\n\\"
\\"Language-Team: LANGUAGE <LL@li.org>\\\\n\\"
\\"Language:\\\\n\\"
\\"MIME-Version: 1.0\\\\n\\"
\\"Content-Type: text/plain; charset=CHARSET\\\\n\\"
\\"Content-Transfer-Encoding: 8bit\\\\n\\"
"
`;

exports[`translation catalog initialization should honor the version option 1`] = `
"# SOME DESCRIPTIVE TITLE
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
# 
#, fuzzy
msgid \\"\\"
msgstr \\"\\"
\\"Project-Id-Version: PACKAGE 23.4.89\\\\n\\"
\\"Report-Msgid-Bugs-To: MSGID_BUGS_ADDRESS\\\\n\\"
\\"POT-Creation-Date: 2020-04-23 08:50+0300\\\\n\\"
\\"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\\\\n\\"
\\"Last-Translator: FULL NAME <EMAIL@ADDRESS>\\\\n\\"
\\"Language-Team: LANGUAGE <LL@li.org>\\\\n\\"
\\"Language:\\\\n\\"
\\"MIME-Version: 1.0\\\\n\\"
\\"Content-Type: text/plain; charset=CHARSET\\\\n\\"
\\"Content-Transfer-Encoding: 8bit\\\\n\\"
"
`;
